Adapt "spacewalk-proxy" code to be Python 2/3 compatible

meaksh opened the pull request:
## What does this PR change?

This PR migrates "spacewalk-proxy" code to be Python 2/3 compatible.

Some notes:
- General fixes addressed by `2to3`.
- Fixed `xmlrpclib`, `urlparse` and `cPickle` import issues.
- Made the spec file to build package for Python3 on SLE15.

## Links

Tracks https://github.com/SUSE/salt-board/issues/62

- [x] **DONE**

isbm commented:
No unicode literals?

isbm commented on proxy/proxy/broker/rhnRepository.py:
```diff
@@ -462,8 +473,8 @@ def cache(stringObject, directory, filename, version):
             fd = os.open(tempfile, os.O_WRONLY | os.O_CREAT | os.O_EXCL,
-                         0644)
-        except OSError, e:
+                         0o644)
```
Since you're touching this already, please put back this silly newline to the previous statement.

isbm commented on proxy/proxy/rhnAuthCacheClient.py:
```diff
@@ -186,9 +191,9 @@ def __str__(self):
-    print 'Cached object s["blah"] = %s' % str(s['blah'])
-    print s.has_key("asdfrasdf")
+    print('Cached object s["1234"] = %s' % str(s['1234']))
+    print('Cached object s["blah"] = %s' % str(s['blah']))
```
I would use format instead.

isbm commented on proxy/proxy/spacewalk-proxy.spec:
```diff
@@ -224,11 +231,19 @@ Conflicts:      rhns-satellite-tools
 
+# Fixing shebang for Python 3
+%if 0%{?build_py3}
+for i in `find . -type f`;
```
Backticks aren't best here. Better:
```bash
for i in $(find . -type-f);
```
...instead.

meaksh commented:
@isbm - I've addressed your feedback and pushed an extra fix for an issue with the `raise` parameters on Python3.

BTW there is no `unicode_literals` here since the code doesn't contain `unicode` nor `u""` at all.

isbm commented on proxy/proxy/rhnAuthCacheClient.py:
```diff
@@ -186,9 +191,9 @@ def __str__(self):
-    print 'Cached object s["1234"] = %s' % str(s['1234'])
-    print 'Cached object s["blah"] = %s' % str(s['blah'])
-    print s.has_key("asdfrasdf")
+    print('Cached object s["1234"] = {}'.format(str(s['1234'])))
```
No need to `str()` this thing then. In Python3 bytes will be rendered as bytes repr, and string is just unicode string.

isbm commented on proxy/proxy/rhnAuthCacheClient.py:
```diff
@@ -186,9 +191,9 @@ def __str__(self):
-    print s.has_key("asdfrasdf")
+    print('Cached object s["1234"] = {}'.format(str(s['1234'])))
+    print('Cached object s["blah"] = {}'.format(str(s['blah'])))
+    print("asdfrasdf" in s)
```
?

meaksh replied:
That code is part of a old test, which seems is never executed during normal operation.

isbm commented:
> BTW there is no unicode_literals here since the code doesn't contain unicode nor u"" at all.

Sure it does. If you use `unicode_literals`, then _all_ the embedded strings of Py2 code becomes unicode, like in Py3. But now they perform differently and chances are this approach will return you different type. You don't want that.

brejoc commented:
Btw, if you add `unicode_literals` and keep the string casting, this will fail with Umlauts.

meaksh commented:
@isbm - After taking a look a the proxy code, AFAIU there is no need to use `unicode_literals` here.

I think this is ready for final review. 

Of course, once we have all the spacewalk python3 packages ready, then we need to enable sumaform to run the testsuite on top of a SUSE Manager server using Python3 (SLE12SP3 or SLE15) and see what are the cucumber results. Currently most of the spacewalk python backend code has no tests at all.
